text/0075-fast-sync.md

```diff
@@ -0,0 +1,155 @@
+
+[guide-level-explanation]: #guide-level-explanation
+
+Fast syncing nodes use a trusted hash to download the database in a
+secure manner. The trusted hash enables secure download of the latest
+block header. That header contains the hash of the database state. Using
+that database hash, nodes download the database state in a
+BitTorent-like manner. After that nodes execute blocks. When they have
+caught up to the current *era* they run the consensus algorithm.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The following `rust` pseudo-code captures the fast-sync logic:
+
+```rust
+pub async fn fast_sync(trusted_hash: Digest, chainspec: &Chainspec) -> BlockHeader {
+    let trusted_header: BlockHeader = fetch_header_by_hash(trusted_hash).await;
+
+    // Get the initial validators
+    let mut parent_hash: Digest = trusted_header.parent_hash;
+    let mut validators: BTreeMap<PubKey, Weight> = loop {
+        let header: BlockHeader = fetch_header_by_hash(parent_hash).await;
+        if let Some(validators) = header.next_era_validator_weights() {
+            break validators;
+        }
+        parent_hash = header.parent_hash;
+    };
+
+    // Get the current header
+    let mut current_header: BlockHeader = trusted_header;
+    while !is_current_era(current_header) {
+        let next_height: u64 = current_header.height + 1;
+        if let Some(next_header) = fetch_header_by_height(validators, next_height).await {
+            current_header = next_header;
+        }
+        if let Some(new_validators) = current_header.next_era_validators_weights() {
+            validators = new_validators;
+        }
+    }
+
+    // Get old blocks so we have deploy hashes to avoid replay attacks
+    let mut walkback_header: BlockHeader = current_header.clone();
+    while walkback_header.timestamp + chainspec.deploy_config.max_ttl > Timestamp::now() {
+        walkback_header = fetch_block_by_hash(walkback_header.parent_hash)
+            .await
+            .header;
+    }
+
+    // Consensus needs the validators of the last couple of eras to initialize
+    // Get the old block headers so we have the validators from the switch-blocks
+    let earliest_era_needed_by_consensus: u64 = current_header.era_id - 2;
+    while walkback_header.era_id > earliest_era_needed_by_consensus {
+        walkback_header = fetch_header_by_hash(walkback_header.parent_hash).await;
+    }
+
+    // Download the block chain database (aka the trie-store)
+    let mut outstanding_trie_hashes: Vec<Digest> = vec![current_header.state_root];
+    while let Some(trie_hash) = outstanding_trie_hashes.pop() {
+        let trie: Trie<Key, StoredValue> = fetch_trie(trie_hash).await;
+        let new_outstanding_trie_hashes =
+            put_trie_and_find_missing_descendant_trie_hashes(trie).await;
+        outstanding_trie_hashes.extend(new_outstanding_trie_hashes);
+    }
+
+    // Download and execute blocks to current
```

Comment:
Maybe add a comment why we're doing this: Downloading the trie store could have taken a long time, so the current `current_header` might not be so current anymore.

Reply:
Done in https://github.com/casper-network/ceps/pull/75/commits/a43ffc357c3c606c0cf7cfe65f5150c1bb408b46.